### PR TITLE
cobalt: Implement zero-copy in-process image transfer path

### PR DESCRIPTION
**Pull request by jasonzhangxx**
Introduce an in-process path for transferring decoded images from the
client to the GPU service.

By bypassing the serialization and deserialization of pixel data when
running in a single-process environment, this optimization reduces
CPU overhead and memory bandwidth. The implementation ensures that
decoded image memory is preserved through the transfer using a closure
runner until the GPU texture upload is finalized.

Bug: 545866687

### CODE REVIEW

File: cc/paint/image_transfer_cache_entry.cc [hunk 1]
```diff
@@ -782,6 +843,94 @@ bool ServiceImageTransferCacheEntry::Deserialize(
...
+    }
+  }
+
+  size_ = image_->textureSize();
+  if (gainmap_image_) {
+    size_ += gainmap_image_->textureSize();
+  }
+
+  return true;
+}
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/paint/image_transfer_cache_entry.cc:1020]: Ack

File: cc/paint/image_transfer_cache_entry.h [hunk 2]
```diff
@@ -96,6 +99,13 @@ class CC_PAINT_EXPORT ClientImageTransferCacheEntry final
   // ClientTransferCacheEntry implementation:
   uint32_t SerializedSize() const final;
   bool Serialize(base::span<uint8_t> data) const final;
+#if BUILDFLAG(IS_COBALT)
+  uint32_t SerializedSizeInProcess() const;
+  bool SerializeInProcess(
+      base::span<uint8_t> data,
+      base::ScopedClosureRunner ref_holder =
+          base::ScopedClosureRunner()) const;
+#endif
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/paint/image_transfer_cache_entry.h]: RasterImplementation::Flush() function will wait until GPU thread has completely processed all pending commands in the transfer cache.

File: cc/paint/image_transfer_cache_entry.cc [hunk 3]
```diff
@@ -34,6 +37,23 @@
...
+  sk_sp<SkColorSpace> target_color_space;
+  bool needs_mips = false;
+
+  // Holds a type-erased keepalive closure (capturing ImageData) to ensure the
+  // underlying decoded pixel buffers remain pinned in RAM and are not unlocked
+  // or deleted until this payload is destroyed on the GPU thread after texture
+  // upload.
+  base::ScopedClosureRunner ref_holder;
+};
+#endif
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/paint/image_transfer_cache_entry.cc]: This cannot prevent memory leak if command buffer is sent to GPU but GPU process doesn't execute the command.

File: cc/paint/image_transfer_cache_entry.cc [hunk 4]
```diff
@@ -526,6 +546,36 @@ uint32_t ClientImageTransferCacheEntry::Id() const {
...
+  payload->needs_mips = needs_mips_;
+  payload->hdr_metadata = hdr_metadata_;
+  payload->target_color_space = target_color_space_;
+  payload->ref_holder = std::move(ref_holder);
+
+  *reinterpret_cast<uintptr_t*>(data.data()) =
+      reinterpret_cast<uintptr_t>(payload);
+  return true;
+}
+#endif
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/paint/image_transfer_cache_entry.cc]: This cannot prevent memory leak if command buffer is sent to GPU but GPU process doesn't execute the command.

File: cc/tiles/gpu_image_decode_cache.cc [hunk 5]
```diff
@@ -2386,11 +2387,35 @@ void GpuImageDecodeCache::InsertTransferCacheEntry(
...
-        UNSAFE_TODO(base::span(static_cast<uint8_t*>(data), size)));
+    bool succeeded = false;
+#if BUILDFLAG(IS_COBALT)
+    if (use_in_process_transfer) {
+      succeeded = image_entry.SerializeInProcess(
+          UNSAFE_TODO(base::span(static_cast<uint8_t*>(data), size)),
+          base::ScopedClosureRunner(
+              base::DoNothingWithBoundArgs(base::WrapRefCounted(image_data))));
+    } else
+#endif
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/tiles/gpu_image_decode_cache.cc]: Updated.

File: cc/tiles/gpu_image_decode_cache.cc [hunk 6]
```diff
@@ -2386,11 +2415,55 @@ void GpuImageDecodeCache::InsertTransferCacheEntry(
...
+                      FROM_HERE,
+                      base::BindOnce(
+                          [](base::WeakPtr<GpuImageDecodeCache> cache,
+                             scoped_refptr<ImageData> image_data) {
+                            if (cache) {
+                              cache->OnInProcessImageTransferCompleted(
+                                  std::move(image_data));
+                            } else if (image_data) {
+                              image_data->decode.ref_count--;
+                            }
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/tiles/gpu_image_decode_cache.cc]: Updated

File: cc/tiles/gpu_image_decode_cache.cc [hunk 7]
```diff
@@ -2260,6 +2271,24 @@ void GpuImageDecodeCache::OwnershipChanged(const DrawImage& draw_image,
...
   }
 
+#if BUILDFLAG(IS_COBALT)
+  OwnershipChanged(image_data);
+}
+
+void GpuImageDecodeCache::OwnershipChanged(ImageData* image_data) {
+  bool has_any_refs =
+      image_data->upload.ref_count > 0 || image_data->decode.ref_count > 0;
+#endif
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/tiles/gpu_image_decode_cache.cc]: Updated

File: cc/paint/image_transfer_cache_entry.cc [hunk 8]
```diff
@@ -526,6 +605,43 @@ uint32_t ClientImageTransferCacheEntry::Id() const {
...
 
+#if BUILDFLAG(IS_COBALT)
+uint32_t ClientImageTransferCacheEntry::SerializedSizeInProcess() const {
+  return sizeof(uintptr_t);
+}
+
+bool ClientImageTransferCacheEntry::SerializeInProcess(
+    base::span<uint8_t> data,
+    base::ScopedClosureRunner unref_runner) const {
+  DCHECK(!IsYuv());
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/paint/image_transfer_cache_entry.cc]: It's not needed.

File: cc/paint/image_transfer_cache_entry.cc [hunk 9]
```diff
@@ -526,6 +605,43 @@ uint32_t ClientImageTransferCacheEntry::Id() const {
...
+  }
+  payload->needs_mips = needs_mips_;
+  payload->hdr_metadata = hdr_metadata_;
+  payload->target_color_space = target_color_space_;
+  payload->unref_runner = std::move(unref_runner);
+
+  InProcessImageTransferCachePayloadRegistry::GetInstance().Register(payload);
+
+  *reinterpret_cast<uintptr_t*>(data.data()) =
+      reinterpret_cast<uintptr_t>(payload);
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/paint/image_transfer_cache_entry.cc]: Done

File: cc/paint/image_transfer_cache_entry.cc [hunk 10]
```diff
@@ -782,6 +909,99 @@ bool ServiceImageTransferCacheEntry::Deserialize(
...
+#if BUILDFLAG(IS_COBALT)
+bool ServiceImageTransferCacheEntry::DeserializeInProcess(
+    GrDirectContext* gr_context,
+    base::span<const uint8_t> data) {
+  DCHECK_EQ(data.size(), sizeof(uintptr_t));
+  if (data.size() != sizeof(uintptr_t)) {
+    return false;
+  }
+  auto* raw_payload = reinterpret_cast<InProcessImageTransferCachePayload*>(
+      *reinterpret_cast<const uintptr_t*>(data.data()));
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/paint/image_transfer_cache_entry.cc]: Done

File: cc/paint/image_transfer_cache_entry.cc [hunk 11]
```diff
@@ -782,6 +909,99 @@ bool ServiceImageTransferCacheEntry::Deserialize(
...
+  if (!img) {
+    return nullptr;
+  }
+  if (gr_context) {
+    img = SkImages::TextureFromImage(
+        gr_context, img,
+        mip_mapped_for_upload ? skgpu::Mipmapped::kYes : skgpu::Mipmapped::kNo,
+        skgpu::Budgeted::kNo);
+    SkImages::GetBackendTextureFromImage(img, /*outTexture=*/nullptr,
+                                        /*flushPendingGrContextIO=*/true);
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/paint/image_transfer_cache_entry.cc]: The code has been updated.

File: cc/tiles/gpu_image_decode_cache.cc [hunk 12]
```diff
@@ -2232,6 +2251,114 @@ void GpuImageDecodeCache::UnrefImageInternal(const DrawImage& draw_image,
   }
 }
 
+#if BUILDFLAG(IS_COBALT)
+// GpuImageDecodeCache::OwnershipChanged(const DrawImage&, ImageData*) is 
+// identical to the chromium version but is splited into two functions.
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/tiles/gpu_image_decode_cache.cc]: Done

File: cc/tiles/gpu_image_decode_cache.cc [hunk 13]
```diff
@@ -2386,11 +2514,53 @@ void GpuImageDecodeCache::InsertTransferCacheEntry(
...
   DCHECK(image_data);
   uint32_t size = image_entry.SerializedSize();
+
+#if BUILDFLAG(IS_COBALT)
+  bool use_in_process_transfer = false;
+  // YUV images are not used in Cobalt, so we currently only support
+  // bypassing serialization for RGBA images.
+  if (base::FeatureList::IsEnabled(
+          features::kCobaltInProcessImageTransferCache) &&
+      !image_entry.IsYuv()) {
```
[the remark this answers is not in the document]
Comment (jasonzhangxx) [reply cc/tiles/gpu_image_decode_cache.cc]: Updated.

**Comment by aelzeiny** [comment]
Tested on RDK

image

b/553484886

File: gpu/command_buffer/client/shared_memory_limits.h [hunk 14]
```diff
@@ -34,6 +39,17 @@ struct SharedMemoryLimits {
...
 #endif
+
+#if BUILDFLAG(IS_COBALT)
+    // When in-process image transfer is enabled, decoded images bypass
+    // transfer cache serialization and no longer use MappedMemoryManager.
+    // MappedMemoryManager is therefore only used for lightweight operations.
+    // Reducing the chunk size to 64KB avoids wasting shared memory.
+    if (base::FeatureList::IsEnabled(
+            base::features::kCobaltInProcessImageTransferCache)) {
+      mapped_memory_chunk_size = 64 * 1024;
```
Comment (sherryzy) [inline gpu/command_buffer/client/shared_memory_limits.h:50]: For this mapped memory chunk size, where does the 64k value originate? Do we know if we can decrease this value further, or if this number will cause any performance issues?

**Review by sherryzy — COMMENTED** [review]
Should we add unit tests for these functions? Additionally, perhaps we can upstream this change once the single-process flag is enabled in Chrome.

File: gpu/command_buffer/client/shared_memory_limits.h (hunk 14 above)
Comment (jasonzhangxx) [reply gpu/command_buffer/client/shared_memory_limits.h:50]: 64k is the `min_transfer_buffer_size` below. When kCobaltInProcessImageTransferCache is enabled, mapped memory is used for font transfer cache (3KB allocation) and 4 bytes for process payload transferring. 16k or 32k can also work. We could just set it to `min_transfer_buffer_size`

**Comment by jasonzhangxx** [comment]
> Should we add unit tests for these functions? Additionally, perhaps we can upstream this change once the single-process flag is enabled in Chrome.

I don't think we can upstream the change as is. To upstream the change, we might need refactor the code to handle all edge cases, like webgl, etc, and of course will need unit tests. It might be challenge to add unit test for this modification.

**Review by sherryzy — APPROVED** [review]

**Comment by jasonzhangxx** [comment]
Created b/555332425 for adding tests.
